src/analyses/variable-sensitivity/full_array_abstract_object.cpp

```diff
@@ -272,113 +270,137 @@ abstract_object_pointert full_array_abstract_objectt::write_element(
       environment, ns, stack, expr, value, merging_write);
   }
 
+  if(!stack.empty())
```

Comment:
It's not entirely clear to me why this is needed.  Can't you just delegate the write to the relevant `abstract_object` and then update the relevant entry with whatever it returns?

Reply:
No, because once we get to the end of the stack we need to worry about whether it's a merging write or not. Higher up the stack, we don't need to do that. 

We have a similar decision point in `full_struct_abstract_objectt`, and I assume we'll have one in our union representation when we fill that out. When we get there, it's entirely possible we'll be able to collapse the implementations.